File: services/python/edtech_service/stores/idempotency_store.py
from __future__ import annotations
from datetime import datetime, timedelta, timezone
from typing import Dict


class IdempotencyStore:
    def __init__(self) -> None:
        self._store: Dict[str, datetime] = {}

    def _now(self) -> datetime:
        return datetime.now(timezone.utc)

    def cleanup(self) -> None:
        now = self._now()
        expired = [k for k, exp in self._store.items() if exp <= now]
        for k in expired:
            self._store.pop(k, None)

    def has(self, key: str) -> bool:
        self.cleanup()
        exp = self._store.get(key)
        return bool(exp and exp > self._now())

    def mark(self, key: str, ttl_seconds: int = 300) -> None:
        self._store[key] = self._now() + timedelta(seconds=ttl_seconds)
        self.cleanup()

    def check_and_mark(self, key: str, ttl_seconds: int = 300) -> bool:
        if self.has(key):
            return False
        self.mark(key, ttl_seconds)
        return True

    def clear(self) -> None:
        self._store.clear()


store = IdempotencyStore()

File: services/node/src/stores/idempotencyStore.js
// Simple in-memory idempotency store with TTL and max size
const DEFAULT_TTL_MS = 24 * 60 * 60 * 1000; // 24h
const MAX_SIZE = 2000;

class IdempotencyStore {
  constructor() {
    this.map = new Map(); // key -> expiresAt
  }

  cleanup(now = Date.now()) {
    for (const [k, exp] of this.map.entries()) {
      if (exp <= now) this.map.delete(k);
    }
    // Trim oldest if over capacity
    while (this.map.size > MAX_SIZE) {
      const firstKey = this.map.keys().next().value;
      this.map.delete(firstKey);
    }
  }

  has(key, now = Date.now()) {
    const exp = this.map.get(key);
    if (!exp) return false;
    if (exp <= now) {
      this.map.delete(key);
      return false;
    }
    return true;
  }

  mark(key, ttlMs = DEFAULT_TTL_MS, now = Date.now()) {
    const expiresAt = now + ttlMs;
    this.map.set(key, expiresAt);
    this.cleanup(now);
  }

  checkAndMark(key, ttlMs = DEFAULT_TTL_MS, now = Date.now()) {
    if (this.has(key, now)) return false;
    this.mark(key, ttlMs, now);
    return true;
  }

  clear() {
    this.map.clear();
  }
}

module.exports = new IdempotencyStore();
